app/controllers/campaigns_controller.rb

```diff
@@ -50,6 +50,39 @@ def associate_seller_distributor
     json_response(campaign_json)
   end
 
+  # POST /campaigns/:id/gift_card
+  def generate_campaign_gift_cards
+    request_params = params.permit(:gift_cards => [:distributor_id, :gift_card_amount, :seller_id])
+    gift_cards = request_params['gift_cards']
+    set_campaign
+
+    total_amount_to_allocate = 0
+    gift_cards.each do |gift_card|
+      total_amount_to_allocate = total_amount_to_allocate + gift_card['gift_card_amount']
+    end
+
+    unless @campaign.amount_unallocated > total_amount_to_allocate
+      raise InvalidLineItem, "Request amount exceeds unallocated amount in campaign. Unallocated amount: #{
+        @campaign.amount_unallocated
+      }"
+    end
+
+    gift_cards.each do |gift_card|
+      WebhookManager::GiftCardCreator.call(
+        {
+          amount: gift_card['gift_card_amount'],
+          single_use: true,
+          project_id: @campaign.project_id,
+          campaign_id: @campaign.id,
+          distributor_id: gift_card['distributor_id'],
+          seller_id: gift_card['seller_id'],
+        }
+      )
+    end
+
+    json_response({'unallocated_amount': @campaign.amount_unallocated})
```

Comment:
Got a bunch of real nitpicky comments you can take or leave, just thought I'd put them all in one comment 😂 

- We usually do the permitting / requiring in a separate function to be called, also I feel like you might want to explicitely not just permit, but **require** the `gift_cards` param here
- You can add your function to the allow-list of `before_actions` and `set_campaign` will be invoked automatically for you
- If you wanna be real fancy you can do a `reduce` for your sum, totally optional though!
- I'm not sure if rails caches the value for `@campaign.amount_unallocated` -- that calls some underlying DB queries, so I'd want to avoid calling it multiple times per API hit if possible. If it does, I'd recommend saving this to a local variable to save time/memory.
- I think `InvalidParameter` is probably a better exception, I think that one was writteexplicitlyly for `line_items` on the `/charges` endpoint... but maybe we're not using it that way anymore anyway, and they take the same path in `exception_handler.rb`
- Maybe return status code `:created` to let the caller know everything went well? And that this is just the remainder left?

Reply:
thanks for the suggestions. i updated the code to reflect them and pinged @justintmckibben about a re-stamp

Reply:
Awesome man, looks really good now, hope these were helpful!